text/0003-hash-indexed-object-store.md

```diff
@@ -0,0 +1,749 @@
+
+_NOTE: hash type is hard coded to simplify the example._
+
+EXAMPLE: `Block` type implementation.
+
+```rust
+pub struct BlockHeader {
+    parent_hash: ObjectHash,
+    root_state_hash: ObjectHash,
+    deploys: Vec<ObjectHash>,
+    era: u64,
+    proofs: Vec<Signature>,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+enum BlockHeaderIter<'a> {
+    Parent(&'a Block),
+    RootState(&'a Block),
+    Deploys(&'a Block, usize),
+    Done,
+}
+
+impl<'a> OutIterable for &'a BlockHeader {
+    type OutIter = BlockHeaderIter<'a>;
+
+    fn iter_outgoing(self) -> Self::OutIter {
+        BlockHeaderIter::Parent(self)
+    }
+}
+
+impl<'a> Iterator for BlockHeaderIter<'a> {
+    type Item = ObjectHash;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut rv = None;
+
+        *self = match *self {
+            BlockHeaderIter::Parent(block) => {
+                rv = block.parent_hash.clone();
+                BlockHeaderIter::RootState(block)
+            }
+            BlockHeaderIter::RootState(block) => {
+                rv = block.root_state.clone();
+                BlockHeaderIter::Deploys(block, 0)
+            }
+            BlockHeaderIter::Deploys(block, idx) if idx < block.deploys.len() => {
+                rv = block.deploys[idx].clone();
+                BlockHeaderIter::Deploys(block, idx + 1)
+            }
+            BlockHeaderIter::Deploys(block, idx)
+            | BlockHeaderIter::Done => BlockHeaderIter::Done,
+        };
+
+        rv
+    }
+}
+```
+
+_NOTE: The above logic is necessary to store and send arbitrary objects using this proposed model; the following logic describes optional convenience methods implemented on `Object` to improve the ergonomics for component developers._
+
+```rust
+impl Object {
+    fn as_block(&self) -> Option<&Block> {
+        match self {
+            Object::Block(ref block) => Some(block),
+            _ => None
+        }
+    }
+
+    // ...
+}
+```
+
+### Storage implementation
+
+Using this proposed model, the storage component decomplects into a single `ObjectStore`, with a simplified set of effects.
+
+```rust
+/// Retrieve multiple objects.
+// This method always uses the heap, since a single is already quite large.
+// Does not return any dependencies, may return less objects than requested.
+pub async fn get_objects(self, object_hash: Vec<ObjectHash>) -> HashMap<ObjectHash, ImmutableObject>;
+
+/// Retrieve multiple objects, including their dependencies.
+///
+/// If specified, stops at a specific depth. Returns `None` if any requested
+/// object was incomplete until `max_depth`.
+pub async fn get_objects_with_dependencies(self, object_hash: Vec<ObjectHash>, max_depth: Option<NonZeroUsize>) -> Option<HashMap<ObjectHash, ImmutableObject>>>;
+
+/// Store multiple objects.
+///
+/// Objects are stored in the reverse order given, so dependencies should be
+/// listed behind the actual object.
+pub async fn put_objects(self, objects: Vec<ImmutableObject>);
```

Comment:
No, I'd like to avoid this. Having "extra" objects lying around should never be harmful, at worst they will get GC'd at a later point in time. Note that there is no way to remove objects =).

If being able to cancel the put due to some condition, I would assume it's better to batch them or even put them one-by-one.

Reply:
> If being able to cancel the put due to some condition, I would assume it's better to batch them or even put them one-by-one.

And indeed, write operations on the trie do create a changelist of new trie nodes which is only written at the very end.  

On a slightly higher level, currently if we are writing many values to trie at once, they are inserted one by one, but some day we will implement a batched write mechanism - it just hasn't been a priority as of yet.

As I mentioned before, I would be comfortable moving to a transaction-less model so long as there was a good GC story.